### DISCUSSION TITLE
Show error to user in action hook

### DISCUSSION THREAD

**Discussion by alexleboucher** [directus/directus#12098] [Archive]
Hello !
Is it possible to show a message like an error to the user in an action hook ? Like the exceptions that we can throw in the filters hook or just an alert maybe I don't know.
Thank you !

**Comment by rijkvanzanten** [comment]
No. Action hooks fire async, which means that the request that original triggered the action can already be done by the time your action is completed. Filter hooks on the other hand fire "synchronous"[^1] so you can modify the payload (or throw errors) before the request is finished and sent to the end user.

[^1]: Still using asynchronous javascript, but it is awaited to allow you to do modifications to the request, rather than just respond to it

**Reply by alexleboucher** [reply]
Thank you for the answer ! But if we encounter an error in the action, like if we call an API and the API throws an error, is there a way to inform the user that an error was encountered ? Not to block or revert the action but only to inform the user.

**Reply by rijkvanzanten** [reply]
If you want to prevent the original operation from finishing when an error happens, you will have to use a filter. If you don't really care about the original operation, and just want to notify the user, you could create a [notification](https://docs.directus.io/reference/system/notifications/) through the NotificationsService. This will show up in the user's module bar (in the bottom left), and send an email if email notifications are enabled. (You could also send an email directly through the MailService, or any other third party means you'd like)

**Reply by hillport** [reply]
Hello 👋,
I'm sorry to bother you again with the same kind of question. I've used the notification system like you suggested previously but it doesn't alert the user directly. 
Is there some way to return a message from a hook to simply inform the user something has been done ?  Like the dialog/toast that appears on the bottom right corner of the screen when an element is updated for instance.
Thanks :)

**Reply by azrikahar** [reply]
> Is there some way to return a message from a hook to simply inform the user something has been done ? Like the dialog/toast that appears on the bottom right corner of the screen when an element is updated for instance.

There's currently no way to do this, but it sounds like a good idea for WebSockets!

**Reply by hillport** [reply]
Ok good to know. Thanks for the quick reply !

**Reply by flevi29** [reply]
Being able to throw an exception that would show up in the admin panel would be great for me too. I can throw an error in `action` but it only shows in logs. I need the ID of a created record, so I need to use `action` instead of `filter`.

**Reply by rijkvanzanten** [reply]
> I need the ID of a created record, so I need to use action instead of filter.

This is a bit of a chicken egg problem then right? It sounds like you want to throw an exception when an item is created, but you need the ID of the item in order to do so. The ID of the item is only generated when the item is saved to the database 🤔

**Reply by 3zzy** [reply]
Wondering if there's a solution to this yet? `files.upload` is only an action, not a filter so can't really show an error and delete the file at the same time.

**Reply by rijkvanzanten** [reply]
@3zzy Are you asking specifically on how to show a custom error message when a new file is being uploaded?

**Reply by 3zzy** [reply]
@rijkvanzanten I guess I can show the error message but then I can't delete the file in the same action thats the issue. And if I delete the file, the error is only logged but not shown to the user since actions are async. So Ideally, I want to show the error to the user and also delete the file in the same `action` if thats possible?

**Reply by rijkvanzanten** [reply]
@3zzy Gotcha. As of right now, there's no "filter" (which is the one that allows you to cancel operations) during the initial upload of a file, due to the way files are streamed to the storage adapter (this is actually something we're currently optimizing). The best next thing is to auto-delete the file after the upload is already done, and then notify the user asynchronously.

**Reply by 3zzy** [reply]
@rijkvanzanten "notify the user asynchronously" - how do you mean, could you please provide an example? Thanks!

**Reply by rijkvanzanten** [reply]
@3zzy I was thinking sending them a notification, or an email or something along those lines. Certainly not as elegant as cancelling the original save outright, but better than nothing. That being said, file uploads always happen in isolation, so you _could_ perform this level of validation when that previously uploaded is being attached to the record. 🤔

**Reply by tiaan-va** [reply]
> @3zzy I was thinking sending them a notification, or an email or something along those lines. Certainly not as elegant as cancelling the original save outright, but better than nothing. That being said, file uploads always happen in isolation, so you _could_ perform this level of validation when that previously uploaded is being attached to the record. 🤔

you were not very helpful in this regard... Also sending them an email to notify the user that the file upload was invalid and thus got deleted in extremely user 'unfriendly'. Like the user would be so confused in the moment and just adds so much hassle to everything. If you guys refuse to add a filter hook to the files.upload event then the least you can do is implement an in-app notification system where we could push a popup notification to the user with a custom message as we please, even if it is not async. Like it could literally be as simple as adding an 'alert("custom message") function which would take no time at all to implement but would be very helpful for a myriad of other reasons too. The fact that we can only send non-tech users actually using the CMS to manage their content alerts via filter hooks is so far behind your CMS competitors.
